apps/settings/main.py

```diff
@@ -1,236 +1,158 @@
-            except:
-                logger.exception("Can't revert failed step {}".format(failed_step))
-                pb.stop()
-                PrettyPrinter("Can't revert failed step '{}'".format(step), i, o, 2)
-                pb.run_in_background()
-            logger.info("Reverting the previous steps")
-            for step in completed_steps:
                 try:
-                    self.revert_step(step)
-                except:
-                    logger.exception("Failed to revert step {}".format(failed_step))
-                    pb.stop()
-                    PrettyPrinter("Failed to revert step '{}'".format(step), i, o, 2)
-                    pb.run_in_background()
-                pb.progress -= progress_per_step
-            sleep(1) # Needed here so that 1) the progressbar goes to 0 2) run_in_background launches the thread before the final stop() call
-            #TODO: add a way to pause the Refresher
-            pb.stop()
-            logger.info("Update failed")
-            PrettyPrinter("Update failed, try again later?", i, o, 3)
+                    bar.message = step.name
+                    step.do()
+                    finished_steps.append(step)
+                    bar.progress = len(finished_steps) / len(self.steps)
+                    bar.refresh()
+                    sleep(0.5)  # so the user has time to read what happens
+                except Exception as e:
+                    if DialogBox("yn", self.i, self.o,
+                                 message="'{}'failed\ncancel update?".format(step.name)).activate():
+                        self.rollback_update(bar, e, finished_steps, step)
+                        return -1
+                    else:
+                        pass
+
+    def rollback_update(self, bar, e, finished_steps, step):
+        logger.error("Error updating step '{}'".format(step.name))
+        logger.exception(e)
+        Printer([step.name, 'update failed'], self.i, self.o)
+        bar.message = "reverting"
+        for i, passed in enumerate(finished_steps):
+            passed.undo()
+            bar.progress = (len(finished_steps) - i) / len(self.steps)
+            bar.refresh()
+            sleep(0.5)
+
+    def create_steps(self, branch="master"):
+        tmp_dir = tempfile.mkdtemp(prefix='zpui')
+        os.chmod(tmp_dir, 0o777)
+        old_cwd = os.getcwd()
+        change_cwd = UpdateStep("opening tmp dir", lambda: os.chdir(tmp_dir), lambda: os.chdir(old_cwd))
+        self.steps.append(change_cwd)
+
+        git_copy = UpdateStep(
+            "copying repo",
+            "git clone {src_dir} {dir}/".format(src_dir=ZPUI_INSTALL_DIR, dir=tmp_dir, branch=branch)
+        )
+        self.steps.append(git_copy)
+
+        git_pull = UpdateStep(
+            "pulling git",
+            "git pull origin {branch} --ff-only".format(dir=tmp_dir, branch=branch)
+        )
+        self.steps.append(git_pull)
+
+        pip_install = UpdateStep("installing deps",
```

Comment:
```
['pip2', 'install', '-r', 'requirements.txt']
Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'
```
Any idea why this could be happening?